### PR TITLE
Get attestation data slot v0.6

### PR DESCRIPTION
**Pull request by cyberbono3**
PR implements [get_attestation_data_slot](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_epoch_start_shard)

[Part of] #2307

---

#Description

1. Created attestation.go and attestation_test.go in beacon-chain/core/helpers
2. Implemented AttestionDataSlot  in attestation.go
3. Generated a new field TargetEpoch for *pb.AttestationData
4. Generated a new field Shard for *pb.Crosslink
3. Copied  get_epoch_committee_count implementation and tests from #2354 
5. Copied get_epoch_start_shard  implementation and test from #2497

#TODO   
Tests

### CODE REVIEW

File: beacon-chain/core/helpers/committee.go [hunk 1]
```diff
@@ -293,41 +304,45 @@ func AttestationParticipants(
...
 	if len(bitfield) != mathutil.CeilDiv8(committeeSize) {
 		return false, fmt.Errorf(
 			"wanted participants bitfield length %d, got: %d",
 			mathutil.CeilDiv8(committeeSize),
 			len(bitfield))
 	}
-	bitLength := len(bitfield) << 3
-	for i := committeeSize; i < bitLength; i++ {
-		set, err := bitutil.CheckBit(bitfield, i)
+
```
Comment (nisdas) [inline beacon-chain/core/helpers/committee.go]: why is this changed ?

Comment (cyberbono3) [reply beacon-chain/core/helpers/committee.go]: I copied it get_epoch_committee_count implementation and tests from #2354
and copied get_epoch_start_shard implementation and test from #2497

File: beacon-chain/core/blocks/block_operations.go [hunk 2]
```diff
@@ -378,6 +378,98 @@ func isSurroundVote(data1 *pb.AttestationData, data2 *pb.AttestationData) bool {
...
+		ProposerIndex: helpers.BeaconProposerIndex(state)
+	}
+	//attestation.data.target_epoch is absent in AttestationData struct fields (types.pb.go)
+	previousEpoch := helpers.PrevEpoch(state)
+	currentEpoch := helpers.CurrentEpoch(state)
+	if data.TargetEpoch != previousEpoch && data.TargetEpoch != currentEpoch  {
+		return nil. fmt.Errorf("TragetEpoch does not march previous previousEpoch or currentEpoch")
+	}
+	
+    type FFGData struct {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: I don't think this struct is needed

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -378,6 +378,98 @@ func isSurroundVote(data1 *pb.AttestationData, data2 *pb.AttestationData) bool {
...
+			RootParam:		 			 state.PreviousJustifiedRoot, 
+			PreviousEpoch:               previousEpoch,
+		}
+		parentCrosslink = state.PreviousCrosslinks[data.Shard]
+		state.PreviousEpochAttestations = append(state.PreviousEpochAttestations, pendingAttestation)
+	}
+	
+
+}
+
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: please fix these spacings

File: beacon-chain/core/helpers/attestation.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+)	
+
+//spec-v0.6
+/*
+def get_attestation_data_slot(state: BeaconState, data: AttestationData) -> Slot:
+committee_count = get_epoch_committee_count(state, data.target_epoch)
+offset = (data.crosslink.shard + SHARD_COUNT - get_epoch_start_shard(state, data.target_epoch)) % SHARD_COUNT
+return get_epoch_start_slot(data.target_epoch) + offset // (committee_count // SLOTS_PER_EPOCH)
+*/
+func AttestionDataSlot(state *pb.BeaconState, data *pb.AttestationData) (uint64)   {
```
Comment (terencechain) [inline beacon-chain/core/helpers/attestation.go]: Please make this function a different PR

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -378,6 +378,98 @@ func isSurroundVote(data1 *pb.AttestationData, data2 *pb.AttestationData) bool {
...
+    # Check FFG data, crosslink data, and signature
+    assert ffg_data == (data.source_epoch, data.source_root, data.target_epoch)
+    assert data.crosslink.epoch == min(data.target_epoch, parent_crosslink.epoch + MAX_EPOCHS_PER_CROSSLINK)
+    assert data.crosslink.parent_root == hash_tree_root(parent_crosslink)
+    assert data.crosslink.data_root == ZERO_HASH  # [to be removed in phase 1]
+    validate_indexed_attestation(state, convert_to_indexed(state, attestation))
+*/
+
+func ProcessBlockAttestations(state *pb.BeaconState, attestation *pb.Attestation) (*pb.BeaconState, error) {
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: Why are we implementing this? The PR says to implement `get_attestation_data_slot` not `process_attestation`

**Comment by cyberbono3** [comment]
@terencechain I fixed issues, in process of writing tests right now
